Apriori algorithm

About Dataset
Source:
Dr Daqing Chen, Director: Public Analytics group. chend '@' lsbu.ac.uk, School of Engineering, London South Bank University, London SE1 0AA, UK.

Data Set Information:
This is a transnational data set which contains all the transactions occurring between 01/12/2010 and 09/12/2011 for a UK-based and registered non-store online retail.The company mainly sells unique all-occasion gifts. Many customers of the company are wholesalers.

Attribute Information:
InvoiceNo: Invoice number. Nominal, a 6-digit integral number uniquely assigned to each transaction. If this code starts with letter 'c', it indicates a cancellation.
StockCode: Product (item) code. Nominal, a 5-digit integral number uniquely assigned to each distinct product.
Description: Product (item) name. Nominal.
Quantity: The quantities of each product (item) per transaction. Numeric.
InvoiceDate: Invice Date and time. Numeric, the day and time when each transaction was generated.
UnitPrice: Unit price. Numeric, Product price per unit in sterling.
CustomerID: Customer number. Nominal, a 5-digit integral number uniquely assigned to each customer.
Country: Country name. Nominal, the name of the country where each customer resides.

Relevant Papers:
The evolution of direct, data and digital marketing, Richard Webber, Journal of Direct, Data and Digital Marketing Practice (2013) 14, 291–309.
Clustering Experiments on Big Transaction Data for Market Segmentation,
Ashishkumar Singh, Grace Rumantir, Annie South, Blair Bethwaite, Proceedings of the 2014 International Conference on Big Data Science and Computing.
A decision-making framework for precision marketing, Zhen You, Yain-Whar Si, Defu Zhang, XiangXiang Zeng, Stephen C.H. Leung c, Tao Li, Expert Systems with Applications, 42 (2015) 3357–3367.

Citation Request:
Daqing Chen, Sai Liang Sain, and Kun Guo, Data mining for the online retail industry: A case study of RFM model-based customer segmentation using data mining, Journal of Database Marketing and Customer Strategy Management, Vol. 19, No. 3, pp. 197–208, 2012 (Published online before print: 27 August 2012. doi: 10.1057/dbm.2012.17).

In [1]:
#import libraries 
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt
from mlxtend.frequent_patterns import apriori,association_rules

df = pd.read_excel('/Users/rameshchidurala/Downloads/Online Retail.xlsx')

In [2]:
# Exploring the columns of the data 
df.columns 

Index(['InvoiceNo', 'StockCode', 'Description', 'Quantity', 'InvoiceDate',
       'UnitPrice', 'CustomerID', 'Country'],
      dtype='object')

In [3]:
# Exploring the different regions of transactions 
df.Country.unique() 

array(['United Kingdom', 'France', 'Australia', 'Netherlands', 'Germany',
       'Norway', 'EIRE', 'Switzerland', 'Spain', 'Poland', 'Portugal',
       'Italy', 'Belgium', 'Lithuania', 'Japan', 'Iceland',
       'Channel Islands', 'Denmark', 'Cyprus', 'Sweden', 'Austria',
       'Israel', 'Finland', 'Bahrain', 'Greece', 'Hong Kong', 'Singapore',
       'Lebanon', 'United Arab Emirates', 'Saudi Arabia',
       'Czech Republic', 'Canada', 'Unspecified', 'Brazil', 'USA',
       'European Community', 'Malta', 'RSA'], dtype=object)

In [4]:
df["Country"].value_counts()

Country
United Kingdom          495478
Germany                   9495
France                    8557
EIRE                      8196
Spain                     2533
Netherlands               2371
Belgium                   2069
Switzerland               2002
Portugal                  1519
Australia                 1259
Norway                    1086
Italy                      803
Channel Islands            758
Finland                    695
Cyprus                     622
Sweden                     462
Unspecified                446
Austria                    401
Denmark                    389
Japan                      358
Poland                     341
Israel                     297
USA                        291
Hong Kong                  288
Singapore                  229
Iceland                    182
Canada                     151
Greece                     146
Malta                      127
United Arab Emirates        68
European Community          61
RSA                         58


In [4]:
#Cleaning the Data

# Stripping extra spaces in the description 
df['Description'] = df['Description'].str.strip() 
  
# Dropping the rows without any invoice number 
df.dropna(axis = 0, subset =['InvoiceNo'], inplace = True) 
df['InvoiceNo'] = df['InvoiceNo'].astype('str') 
  
# Dropping all transactions which were done on credit 
df = df[~df['InvoiceNo'].str.contains('C')] 

In [5]:
#Splitting the data according to the region of transaction

# Transactions done in France 
basket_France = (df[df['Country'] =="France"] 
          .groupby(['InvoiceNo', 'Description'])['Quantity'] 
          .sum().unstack().reset_index().fillna(0) 
          .set_index('InvoiceNo')) 
  
# Transactions done in the United Kingdom 
basket_UK = (df[df['Country'] =="United Kingdom"] 
          .groupby(['InvoiceNo', 'Description'])['Quantity'] 
          .sum().unstack().reset_index().fillna(0) 
          .set_index('InvoiceNo')) 
  
# Transactions done in Portugal 
basket_Por = (df[df['Country'] =="Portugal"] 
          .groupby(['InvoiceNo', 'Description'])['Quantity'] 
          .sum().unstack().reset_index().fillna(0) 
          .set_index('InvoiceNo')) 
  
basket_Sweden = (df[df['Country'] =="Sweden"] 
          .groupby(['InvoiceNo', 'Description'])['Quantity'] 
          .sum().unstack().reset_index().fillna(0) 
          .set_index('InvoiceNo')) 

In [6]:
#Hot encoding the Data

# Defining the hot encoding function to make the data suitable  
# for the concerned libraries 
def hot_encode(x): 
    if(x<= 0): 
        return 0
    if(x>= 1): 
        return 1
  
# Encoding the datasets 
basket_encoded = basket_France.applymap(hot_encode) 
basket_France = basket_encoded 
  
basket_encoded = basket_UK.applymap(hot_encode) 
basket_UK = basket_encoded 
  
basket_encoded = basket_Por.applymap(hot_encode) 
basket_Por = basket_encoded 
  
basket_encoded = basket_Sweden.applymap(hot_encode) 
basket_Sweden = basket_encoded 

/var/folders/kw/dtj9__8n4qv9bxydv0cs61380000gn/T/ipykernel_9394/2214807749.py:12: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  basket_encoded = basket_France.applymap(hot_encode)
/var/folders/kw/dtj9__8n4qv9bxydv0cs61380000gn/T/ipykernel_9394/2214807749.py:15: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  basket_encoded = basket_UK.applymap(hot_encode)
/var/folders/kw/dtj9__8n4qv9bxydv0cs61380000gn/T/ipykernel_9394/2214807749.py:18: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  basket_encoded = basket_Por.applymap(hot_encode)
/var/folders/kw/dtj9__8n4qv9bxydv0cs61380000gn/T/ipykernel_9394/2214807749.py:21: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  basket_encoded = basket_Sweden.applymap(hot_encode)


In [7]:
#Building the models and analyzing the results
#a) France:

# Building the model 
frq_items = apriori(basket_France, min_support = 0.05, use_colnames = True) 
  
# Collecting the inferred rules in a dataframe 
rules = association_rules(frq_items, metric ="lift", min_threshold = 1) 
rules = rules.sort_values(['confidence', 'lift'], ascending =[False, False]) 
rules

/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/mlxtend/frequent_patterns/fpcommon.py:109: DeprecationWarning: DataFrames with non-bool types result in worse computationalperformance and their support might be discontinued in the future.Please use a DataFrame with bool type
  warnings.warn(


antecedents  \
44                        (JUMBO BAG WOODLAND ANIMALS)   
260  (PLASTERS IN TIN CIRCUS PARADE, RED TOADSTOOL ...   
272  (RED TOADSTOOL LED NIGHT LIGHT, PLASTERS IN TI...   
300  (SET/6 RED SPOTTY PAPER CUPS, SET/20 RED RETRO...   
302  (SET/20 RED RETROSPOT PAPER NAPKINS, SET/6 RED...   
..                                                 ...   
36                                           (POSTAGE)   
27                                           (POSTAGE)   
96                                           (POSTAGE)   
226                                          (POSTAGE)   
214                                          (POSTAGE)   

                                           consequents  antecedent support  \
44                                           (POSTAGE)            0.076531   
260                                          (POSTAGE)            0.051020   
272                                          (POSTAGE)            0.053571   
300                    (SET/6 RED SPOTTY PAPER PLATES)            0.102041   
302                      (SET/6 RED SPOTTY PAPER CUPS)            0.102041   
..                                                 ...                 ...   
36                            (JAM MAKING SET PRINTED)            0.765306   
27                   (CIRCUS PARADE CHILDRENS EGG CUP)            0.765306   
96                                     (PARTY BUNTING)            0.765306   
226      (LUNCH BAG WOODLAND, LUNCH BAG RED RETROSPOT)            0.765306   
214  (LUNCH BAG APPLE DESIGN, LUNCH BAG SPACEBOY DE...            0.765306   

     consequent support   support  confidence      lift  leverage  conviction  \
44             0.765306  0.076531    1.000000  1.306667  0.017961         inf   
260            0.765306  0.051020    1.000000  1.306667  0.011974         inf   
272            0.765306  0.053571    1.000000  1.306667  0.012573         inf   
300            0.127551  0.099490    0.975000  7.644000  0.086474   34.897959   
302            0.137755  0.099490    0.975000  7.077778  0.085433   34.489796   
..                  ...       ...         ...       ...       ...         ...   
36             0.053571  0.051020    0.066667  1.244444  0.010022    1.014031   
27             0.056122  0.051020    0.066667  1.187879  0.008070    1.011297   
96             0.056122  0.051020    0.066667  1.187879  0.008070    1.011297   
226            0.056122  0.051020    0.066667  1.187879  0.008070    1.011297   
214            0.063776  0.051020    0.066667  1.045333  0.002213    1.003098   

     zhangs_metric  
44        0.254144  
260       0.247312  
272       0.247978  
300       0.967949  
302       0.956294  
..             ...  
36        0.836957  
27        0.673913  
96        0.673913  
226       0.673913  
214       0.184783  

[348 rows x 10 columns]

From the above output, it can be seen that paper cups and paper and plates are bought together in France. This is because the French have a culture of having a get-together with their friends and family atleast once a week. Also, since the French government has banned the use of plastic in the country, the people have to purchase the paper-based alternatives. 

In [8]:
'''
from mlxtend.frequent_patterns import apriori, association_rules
#b)UK
# Assuming basket_UK is a properly formatted DataFrame where rows are transactions and columns are items with 1s and 0s
frq_items = apriori(basket_UK, min_support=0.01, use_colnames=True)
rules = association_rules(frq_items, metric="lift", min_threshold=1)
rules = rules.sort_values(['confidence', 'lift'], ascending=[False, False])
print(rules.head())
'''

'\nfrom mlxtend.frequent_patterns import apriori, association_rules\n\n# Assuming basket_UK is a properly formatted DataFrame where rows are transactions and columns are items with 1s and 0s\nfrq_items = apriori(basket_UK, min_support=0.01, use_colnames=True)\nrules = association_rules(frq_items, metric="lift", min_threshold=1)\nrules = rules.sort_values([\'confidence\', \'lift\'], ascending=[False, False])\nprint(rules.head())\n'

The kernel is crashing due to a large number of rows, 495478 . So Iam commenting it and running the rules for different countrys.

In [12]:
# C)Portugal
frq_items = apriori(basket_Por, min_support = 0.05, use_colnames = True) 
rules = association_rules(frq_items, metric ="lift", min_threshold = 1) 
rules = rules.sort_values(['confidence', 'lift'], ascending =[False, False]) 
rules

/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/mlxtend/frequent_patterns/fpcommon.py:109: DeprecationWarning: DataFrames with non-bool types result in worse computationalperformance and their support might be discontinued in the future.Please use a DataFrame with bool type
  warnings.warn(


antecedents  \
1170  (SET 12 COLOUR PENCILS DOLLY GIRL)   
1171    (SET 12 COLOUR PENCILS SPACEBOY)   
1172  (SET OF 4 KNICK KNACK TINS LONDON)   
1173  (SET 12 COLOUR PENCILS DOLLY GIRL)   
1174  (SET 12 COLOUR PENCILS DOLLY GIRL)   
...                                  ...   
1056                           (POSTAGE)   
1059                           (POSTAGE)   
1067                           (POSTAGE)   
1070                           (POSTAGE)   
1209                           (POSTAGE)   

                                            consequents  antecedent support  \
1170                   (SET 12 COLOUR PENCILS SPACEBOY)            0.051724   
1171                 (SET 12 COLOUR PENCILS DOLLY GIRL)            0.051724   
1172                 (SET 12 COLOUR PENCILS DOLLY GIRL)            0.051724   
1173                 (SET OF 4 KNICK KNACK TINS LONDON)            0.051724   
1174                (SET OF 4 KNICK KNACK TINS POPPIES)            0.051724   
...                                                 ...                 ...   
1056                         (REGENCY CAKESTAND 3 TIER)            0.517241   
1059                 (RETROSPOT HEART HOT WATER BOTTLE)            0.517241   
1067                 (SET OF 3 CAKE TINS PANTRY DESIGN)            0.517241   
1070                  (SET OF 36 TEATIME PAPER DOILIES)            0.517241   
1209  (CHARLOTTE BAG SUKI DESIGN, 4 TRADITIONAL SPIN...            0.517241   

      consequent support   support  confidence       lift  leverage  \
1170            0.051724  0.051724         1.0  19.333333  0.049049   
1171            0.051724  0.051724         1.0  19.333333  0.049049   
1172            0.051724  0.051724         1.0  19.333333  0.049049   
1173            0.051724  0.051724         1.0  19.333333  0.049049   
1174            0.051724  0.051724         1.0  19.333333  0.049049   
...                  ...       ...         ...        ...       ...   
1056            0.086207  0.051724         0.1   1.160000  0.007134   
1059            0.086207  0.051724         0.1   1.160000  0.007134   
1067            0.086207  0.051724         0.1   1.160000  0.007134   
1070            0.086207  0.051724         0.1   1.160000  0.007134   
1209            0.086207  0.051724         0.1   1.160000  0.007134   

      conviction  zhangs_metric  
1170         inf       1.000000  
1171         inf       1.000000  
1172         inf       1.000000  
1173         inf       1.000000  
1174         inf       1.000000  
...          ...            ...  
1056    1.015326       0.285714  
1059    1.015326       0.285714  
1067    1.015326       0.285714  
1070    1.015326       0.285714  
1209    1.015326       0.285714  

[335164 rows x 10 columns]

On analyzing the association rules for Portuguese transactions, it is observed that Tiffin sets (Knick Knack Tins) and color pencils. These two products typically belong to a primary school going kid. These two products are required by children in school to carry their lunch and for creative work respectively and hence are logically make sense to be paired together.

In [11]:
#d) Sweden:

frq_items = apriori(basket_Sweden, min_support = 0.05, use_colnames = True) 
rules = association_rules(frq_items, metric ="lift", min_threshold = 1) 
rules = rules.sort_values(['confidence', 'lift'], ascending =[False, False]) 
rules

/Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/mlxtend/frequent_patterns/fpcommon.py:109: DeprecationWarning: DataFrames with non-bool types result in worse computationalperformance and their support might be discontinued in the future.Please use a DataFrame with bool type
  warnings.warn(


antecedents  \
0       (PACK OF 72 SKULL CAKE CASES)   
1       (12 PENCILS SMALL TUBE SKULL)   
4      (ASSORTED BOTTLE TOP  MAGNETS)   
5             (36 DOILIES DOLLY GIRL)   
180    (CHILDRENS CUTLERY DOLLY GIRL)   
...                               ...   
25485                       (POSTAGE)   
202                         (POSTAGE)   
384                         (POSTAGE)   
482                         (POSTAGE)   
3571                        (POSTAGE)   

                                             consequents  antecedent support  \
0                          (12 PENCILS SMALL TUBE SKULL)            0.055556   
1                          (PACK OF 72 SKULL CAKE CASES)            0.055556   
4                                (36 DOILIES DOLLY GIRL)            0.055556   
5                         (ASSORTED BOTTLE TOP  MAGNETS)            0.055556   
180                    (CHILDRENS CUTLERY CIRCUS PARADE)            0.055556   
...                                                  ...                 ...   
25485  (CHILDRENS CUTLERY DOLLY GIRL, CHILDRENS CUTLE...            0.611111   
202                           (CUPCAKE LACE PAPER SET 6)            0.611111   
384                      (MINI PLAYING CARDS DOLLY GIRL)            0.611111   
482                 (ROUND SNACK BOXES SET OF4 WOODLAND)            0.611111   
3571   (ROUND SNACK BOXES SET OF 4 FRUITS, ROUND SNAC...            0.611111   

       consequent support   support  confidence       lift  leverage  \
0                0.055556  0.055556    1.000000  18.000000  0.052469   
1                0.055556  0.055556    1.000000  18.000000  0.052469   
4                0.055556  0.055556    1.000000  18.000000  0.052469   
5                0.055556  0.055556    1.000000  18.000000  0.052469   
180              0.055556  0.055556    1.000000  18.000000  0.052469   
...                   ...       ...         ...        ...       ...   
25485            0.055556  0.055556    0.090909   1.636364  0.021605   
202              0.083333  0.055556    0.090909   1.090909  0.004630   
384              0.083333  0.055556    0.090909   1.090909  0.004630   
482              0.083333  0.055556    0.090909   1.090909  0.004630   
3571             0.083333  0.055556    0.090909   1.090909  0.004630   

       conviction  zhangs_metric  
0             inf       1.000000  
1             inf       1.000000  
4             inf       1.000000  
5             inf       1.000000  
180           inf       1.000000  
...           ...            ...  
25485    1.038889       1.000000  
202      1.008333       0.214286  
384      1.008333       0.214286  
482      1.008333       0.214286  
3571     1.008333       0.214286  

[104608 rows x 10 columns]

On analyzing the above rules, it is found that boys’ and girls’ cutlery are paired together. This makes practical sense because when a parent goes shopping for cutlery for his/her children, he/she would want the product to be a little customized according to the kid’s wishes.

Limitations of Apriori Algorithm
Apriori Algorithm can be slow. The main limitation is time required to hold a vast number of candidate sets with much frequent itemsets, low minimum support or large itemsets i.e. it is not an efficient approach for large number of datasets. For example, if there are 10^4 from frequent 1- itemsets, it need to generate more than 10^7 candidates into 2-length which in turn they will be tested and accumulate. Furthermore, to detect frequent pattern in size 100 i.e. v1, v2… v100, it have to generate 2^100 candidate itemsets that yield on costly and wasting of time of candidate generation. So, it will check for many sets from candidate itemsets, also it will scan database many times repeatedly for finding candidate itemsets. Apriori will be very low and inefficiency when memory capacity is limited with large number of transactions.[Source : https://arxiv.org/pdf/1403.3948.pdf]

References:


Chen, D., Sain, S. L., & Guo, K. (2012). Data mining for the online retail industry: A case study of RFM model-based customer segmentation using data mining. Journal of Database Marketing and Customer Strategy Management, 19(3), 197–208. https://doi.org/10.1057/dbm.2012.17

Kaggle. (2021). Online Retail [Data set]. https://www.kaggle.com/datasets/tunguz/online-retail